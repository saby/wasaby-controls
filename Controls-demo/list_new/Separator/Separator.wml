<div class="controlsDemo__wrapper controlDemo_list-new_base">
    <div class="controlsDemo__inline-block" style="margin-right: 20px;">
        <Controls.list:View keyProperty="key" source="{{_viewSource}}" rowSeparatorSize="s"/>
    </div>

    <div class="controlsDemo__inline-block" style="margin-right: 20px;">
        <Controls.list:View keyProperty="key" source="{{_viewSource}}" rowSeparatorSize="l"/>
    </div>
</div>
