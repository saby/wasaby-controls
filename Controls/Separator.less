@import 'Controls/_Separator/_Separator';
