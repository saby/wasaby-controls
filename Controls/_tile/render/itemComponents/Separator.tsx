import * as React from 'react';

import { TInternalProps } from 'UICore/Executor';
import type { RootSeparatorItem } from 'Controls/baseTree';

interface IProps extends TInternalProps {
    item: RootSeparatorItem;
}

function Separator(props: IProps, ref: React.ForwardedRef<HTMLDivElement>) {
    const wrapperClassName = props.item.getWrapperClasses();
    return (
        <div ref={ref} className={wrapperClassName}>
            <div className={'controls-TileView__searchSeparator_line_horizontal'} />
        </div>
    );
}

export default React.forwardRef(Separator);
