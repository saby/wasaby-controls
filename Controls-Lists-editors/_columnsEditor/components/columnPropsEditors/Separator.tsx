import { Control } from 'Controls/Chips';
import { SyntheticEvent } from 'UICommon/Events';
import * as React from 'react';

/**
 * Редактор свойства "Границы" в колонке
 * @param props
 * @constructor
 */
export function ColumnSeparatorEditor(props): JSX.Element {
    const { value, onChange, LayoutComponent, items } = props;
    const separatorFields = React.useMemo(() => {
        return ['right', 'left'];
    }, []);
    const selectedKeys = React.useMemo(() => {
        return separatorFields.filter((field: string) => {
            return value[field] === 'bold';
        });
    }, [separatorFields, value]);
    const onSelectedKeysChanged = React.useCallback(
        (event: SyntheticEvent<Event>, keys: string[]) => {
            const newValue = {};
            separatorFields.forEach((field: string) => {
                const selected = keys.includes(field);
                if (selected) {
                    newValue[field] = 'bold';
                } else {
                    newValue[field] = 's';
                }
            });
            onChange(newValue);
        },
        [separatorFields, onChange, selectedKeys]
    );
    return (
        <LayoutComponent>
            <Control
                items={items}
                viewMode={'filled'}
                contrastBackground={false}
                fontSize={'4xl'}
                onSelectedKeysChanged={onSelectedKeysChanged}
                selectedKeys={selectedKeys}
            />
        </LayoutComponent>
    );
}
