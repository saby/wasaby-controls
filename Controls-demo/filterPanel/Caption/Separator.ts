import { Control, TemplateFunction } from 'UI/Base';
import { IFilterItem } from 'Controls/filter';
import * as Template from 'wml!Controls-demo/filterPanel/Caption/Base';

export default class extends Control {
    protected _template: TemplateFunction = Template;
    protected _filterSource: IFilterItem[] = [];

    protected _beforeMount(): void {
        this._filterSource = [
            {
                name: 'amount',
                editorTemplateName:
                    'Controls/filterPanelExtEditors:NumberRangeEditor',
                resetValue: [],
                value: [],
                caption: '',
            },
        ];
    }

    static _styles: string[] = ['Controls-demo/filterPanel/Index'];
}
