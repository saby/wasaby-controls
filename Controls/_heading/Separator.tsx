/**
 * @kaizen_zone cf38e892-5e45-4941-98a7-87bbb1838d31
 */
import * as React from 'react';
import { IControlProps } from 'Controls/interface';
import { Logger } from 'UI/Utils';
import 'css!Controls/heading';
import { TInternalProps } from 'UICore/Executor';
import { wasabyAttrsToReactDom } from 'UICore/Jsx';

export interface ISeparatorOptions extends IControlProps, TInternalProps {
    separatorStyle?: 'primary' | 'secondary';
    style?: string;
}

/**
 * Разделитель заголовков с поддержкой некоторых стилей отображения.
 *
 * @remark
 * Используется в составе сложных заголовков, состоящих из {@link Controls/heading:Separator}, {@link Controls/heading:Counter} и {@link Controls/heading:Title}.
 * Для одновременной подсветки всех частей сложного заголовка при наведении используйте класс controls-Header_all__clickable на контейнере.
 *
 * Полезные ссылки:
 * * {@link /doc/platform/developmentapl/interface-development/controls/text-and-styles/heading/ руководство разработчика}
 * * {@link https://github.com/saby/wasaby-controls/blob/rc-20.4000Controls-default-theme/variables/_heading.less переменные тем оформления}
 *
 *
 * @class Controls/_heading/Separator
 * @implements Controls/interface:IControl
 * @implements Controls/interface:ICaption
 * @public
 *
 * @demo Controls-demo/Heading/Separators/Index
 */

/*
 * Heading separator with support some display styles. Used as part of complex headings(you can see it in Demo-example)
 * consisting of a <a href="/docs/js/Controls/_heading/?v=3.18.500">header</a>, a <a href="/docs/js/Controls/Button/Separator/?v=3.18.500">button-separator</a> and a <a href="/docs/js/Controls/_heading/Counter/?v=3.18.500">counter</a>.
 *
 * <a href="/materials/DemoStand/app/Controls-demo%2FHeaders%2FstandartDemoHeader">Demo-example</a>.
 *
 * @class Controls/_heading/Separator
 *
 * @public
 * @author Мочалов М.А.
 *
 * @demo Controls-demo/Heading/Separators/Index
 */

export default React.forwardRef(function Separator(
    props: ISeparatorOptions,
    _
): React.ReactElement {
    const [separatorStyle, setSeparatorStyle] = React.useState<string>('');

    React.useEffect(() => {
        setSeparatorStyle(props.style || props.separatorStyle || 'secondary');
        if (props.style !== undefined) {
            Logger.warn(
                'Header.Separator: Используется устаревшая опция style,' +
                    ' нужно использовать separatorStyle',
                this
            );
        }
    }, []);

    const attrs = props.attrs ? wasabyAttrsToReactDom(props.attrs) || {} : {};

    return (
        <div
            {...attrs}
            ref={props.$wasabyRef}
            className={`controls-HeaderSeparator ${attrs.className} ${
                props.className || ''
            }`}
        >
            <div className="controls-HeaderSeparator__wrapper">
                <svg
                    viewBox="0 0 6 10"
                    xmlns="http://www.w3.org/2000/svg"
                    className={`controls-HeaderSeparator__arrow controls-HeaderSeparator__arrow_style-${separatorStyle}`}
                >
                    <polygon points="0,0 2,0 6,5 2,10 0,10 4,5" />
                </svg>
            </div>
        </div>
    );
});

/**
 * @name Controls/_heading/Separator#separatorStyle
 * @cfg {String} Стиль отображения иконки. В теме онлайна есть только один стиль отображения.
 * @variant primary
 * @variant secondary
 * @default secondary
 */

/*
 * @name Controls/_heading/Separator#style
 * @cfg {String} Icon display style. In the online theme has only one display style.
 * @variant primary
 * @variant secondary
 * @default secondary
 */
