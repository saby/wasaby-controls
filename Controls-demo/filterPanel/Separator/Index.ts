import { Control, TemplateFunction } from 'UI/Base';
import * as Template from 'wml!Controls-demo/filterPanel/Separator/Index';

export default class extends Control {
    protected _template: TemplateFunction = Template;
    protected _filterSource: object[] = [];

    protected _beforeMount(): void {
        this._filterSource = [
            {
                caption: '',
                name: 'amount',
                editorTemplateName:
                    'Controls/filterPanelExtEditors:NumberRangeEditor',
                resetValue: [],
                value: [],
            },
        ];
    }

    static _styles: string[] = ['Controls-demo/filterPanel/Index'];
}
