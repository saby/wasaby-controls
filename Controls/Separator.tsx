/**
 * @kaizen_zone 3f785aa8-d36c-4b57-946a-a916e51ded4d
 */
import * as React from 'react';
import { IControlOptions } from 'UI/Base';
import { ICheckableOptions } from 'Controls/interface';
import { wasabyAttrsToReactDom } from 'UICore/Jsx';
import { IWasabyAttributes } from 'UICore/Executor';
import 'css!Controls/Separator';

export interface ISeparatorProps extends IControlOptions, ICheckableOptions {
    style?: 'primary' | 'secondary' | 'unaccented';
    bold?: boolean;
    onValuechanged?: (e: React.MouseEvent<HTMLSpanElement, MouseEvent>, value: boolean) => void;
    attrs?: IWasabyAttributes;
}

/**
 * Кнопка-разделитель с поддержкой различных стилей отображения и жирным шрифтом. Может использоваться как самостоятельно, так и в составе {@link Controls/heading сложных заголовков}.
 *
 * @remark
 * Полезные ссылки:
 * * {@link /materials/DemoStand/app/Controls-demo%2FHeaders%2FstandartDemoHeader демо-пример}
 * * {@link https://github.com/saby/wasaby-controls/blob/rc-20.4000/Controls-default-theme/variables/_toggle.less переменные тем оформления}
 *
 *
 * @class Controls/Separator
 * @extends UI/Base:Control
 *
 * @public
 * @implements Controls/interface:ICheckable
 *
 * @demo Controls-demo/toggle/Separator/Index
 *
 */

/*
 * Button separator with support different display styles and can be bold thickness. Can be used independently or as part of complex headers(you can see it in Demo-example)
 * consisting of a <a href="/docs/js/Controls/Heading/?v=3.18.500">header</a>, a <a href="/docs/js/Controls/Heading/Separator/?v=3.18.500">header-separator</a> and a <a href="/docs/js/Controls/Heading/Counter/?v=3.18.500">counter</a>.
 *
 * <a href="/materials/DemoStand/app/Controls-demo%2FHeaders%2FstandartDemoHeader">Demo-example</a>.
 *
 * @class Controls/Separator
 * @extends UI/Base:Control
 *
 * @public
 * @author Мочалов М.А.
 * @implements Controls/interface:ICheckable
 *
 * @demo Controls-demo/toggle/Separator/Index
 *
 */
const Separator = React.memo(function (props: ISeparatorProps): React.ReactElement {
    const clickHandler = (e: React.MouseEvent<HTMLSpanElement, MouseEvent>) => {
        if (props.onValuechanged) {
            props.onValuechanged(e, !value);
        }
    };
    let icon;
    const { value = false, bold = false } = props;
    if (value) {
        icon = 'icon-' + (bold ? 'MarkCollapseBold ' : 'CollapseLight ');
    } else {
        icon = 'icon-' + (bold ? 'MarkExpandBold ' : 'ExpandLight ');
    }
    const attrs = wasabyAttrsToReactDom(props.attrs || {});
    const attrsClassName =
        `controls_toggle_theme-${props.theme || 'default'} controls-Separator__icon` +
        ' ' +
        (attrs.className || '');
    return (
        <span
            onClick={clickHandler}
            {...attrs}
            className={attrsClassName + ' ' + (props.className || '')}
            data-qa="controls-Separator"
        >
            <span className="controls-Separator__wrapper controls-Separator__wrapper">
                <i
                    className={
                        `controls-icon controls-Separator__icon_style-${
                            props.style || 'secondary'
                        }` + ` ${icon} controls-icon controls-Separator__icon_size`
                    }
                    data-qa={`controls-Separator__${props.value ? 'collapse' : 'expand'}`}
                ></i>
            </span>
        </span>
    );
});

/**
 * @name Controls/Separator#style
 * @cfg {String} Стиль отображения разделителя.
 * @variant secondary
 * @variant unaccented
 * @variant primary
 */

/*
 * @name Controls/Separator#style
 * @cfg {String} Separator display style.
 * @variant secondary
 * @variant additional
 * @variant primary
 */

/**
 * @name Controls/Separator#bold
 * @cfg {Boolean} Определяет толщину двойного разделителя.
 */

/*
 * @name Controls/Separator#bold
 * @cfg {Boolean} Determines the double separator thickness.
 */
export default Separator;
