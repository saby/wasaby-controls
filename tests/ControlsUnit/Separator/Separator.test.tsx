/**
 * @jest-environment jsdom
 */
import * as React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import Separator from 'Controls/Separator';

describe('Controls/Separator', function () {
    let container = null;
    beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
    });
    afterEach(() => {
        unmountComponentAtNode(container);
        container.remove();
        container = null;
    });
    it('default', () => {
        act(() => {
            render(<Separator />, container);
        });
        expect(container).toMatchSnapshot();
    });

    it('value=true', function () {
        act(() => {
            render(<Separator value={true} />, container);
        });
        expect(container).toMatchSnapshot();
    });

    it('bold=true', function () {
        act(() => {
            render(<Separator bold={true} />, container);
        });
        expect(container).toMatchSnapshot();
    });

    it('style=primary', function () {
        act(() => {
            render(<Separator style="primary" />, container);
        });
        expect(container).toMatchSnapshot();
    });
});
