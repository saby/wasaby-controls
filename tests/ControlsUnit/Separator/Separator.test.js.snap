// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls/Separator bold=true 1`] = `
<div>
  <span
    class="controls_toggle_theme-default controls-Separator__icon  "
    data-qa="controls-Separator"
  >
    <span
      class="controls-Separator__wrapper controls-Separator__wrapper"
    >
      <i
        class="controls-icon controls-Separator__icon_style-secondary icon-MarkExpandBold  controls-icon controls-Separator__icon_size"
        data-qa="controls-Separator__expand"
      />
    </span>
  </span>
</div>
`;

exports[`Controls/Separator default 1`] = `
<div>
  <span
    class="controls_toggle_theme-default controls-Separator__icon  "
    data-qa="controls-Separator"
  >
    <span
      class="controls-Separator__wrapper controls-Separator__wrapper"
    >
      <i
        class="controls-icon controls-Separator__icon_style-secondary icon-ExpandLight  controls-icon controls-Separator__icon_size"
        data-qa="controls-Separator__expand"
      />
    </span>
  </span>
</div>
`;

exports[`Controls/Separator style=primary 1`] = `
<div>
  <span
    class="controls_toggle_theme-default controls-Separator__icon  "
    data-qa="controls-Separator"
  >
    <span
      class="controls-Separator__wrapper controls-Separator__wrapper"
    >
      <i
        class="controls-icon controls-Separator__icon_style-primary icon-ExpandLight  controls-icon controls-Separator__icon_size"
        data-qa="controls-Separator__expand"
      />
    </span>
  </span>
</div>
`;

exports[`Controls/Separator value=true 1`] = `
<div>
  <span
    class="controls_toggle_theme-default controls-Separator__icon  "
    data-qa="controls-Separator"
  >
    <span
      class="controls-Separator__wrapper controls-Separator__wrapper"
    >
      <i
        class="controls-icon controls-Separator__icon_style-secondary icon-CollapseLight  controls-icon controls-Separator__icon_size"
        data-qa="controls-Separator__collapse"
      />
    </span>
  </span>
</div>
`;
